src/backend/cuda/kernel/anisotropic_diffusion.hpp

```diff
@@ -0,0 +1,239 @@
+/*******************************************************
+ * Copyright (c) 2017, ArrayFire
+ * All rights reserved.
+ *
+ * This file is distributed under 3-clause BSD license.
+ * The complete license agreement can be obtained at:
+ * http://arrayfire.com/licenses/BSD-3-Clause
+ ********************************************************/
+
+#include <backend.hpp>
+#include <common/dispatch.hpp>
+#include <Param.hpp>
+#include <debug_cuda.hpp>
+#include <math.hpp>
+#include <ops.hpp>
+
+namespace cuda
+{
+namespace kernel
+{
+static const int THREADS_X = 32;
+static const int THREADS_Y = 8;
+
+inline __device__
+int clamp(int value, int low, int high)
+{
+    return max(low, min(value, high));
+}
+
+__forceinline__ __device__
+int index(int x, int y, int dim0, int dim1, int stride0, int stride1)
+{
+    return clamp(x, 0, dim0-1)*stride0 + clamp(y, 0, dim1-1)*stride1;
+}
+
+__device__
+float quad(float value)
+{
+    return 1.0/(1.0+value);
+}
+
+template<af_flux_function fftype>
```

Comment:
Have you seen a significant increase in performance by making this a template argument?

Reply:
Nope, just tested it out and removed all unnecessary templates.